### PR TITLE
net: Refactor message parsing (CNetMessage), adds flexibility

### PR DESCRIPTION
**Pull request by jonasschnelli**
This refactors the code how incoming messages gets handled.
It adds flexibility for adding new transport types (encryption handshake, new message format).

This is a subset of #14032 and a pre-requirement for BIP151.

### CODE REVIEW

File: src/net_message.h [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+/** Maximum length of incoming protocol messages (no message over 4 MB is currently acceptable). */
+static const unsigned int MAX_PROTOCOL_MESSAGE_LENGTH = 4 * 1000 * 1000;
+
+// base class for format agnostic network messages
+class NetMessageBase
+{
+public:
+    CDataStream vRecv; // received message data
+    int64_t nTime;     // time (in microseconds) of message receipt.
```
Comment (scravy) [inline src/net_message.h]: According to https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md these should be prefixed `m_`.

**Comment by promag** [comment]
Any reason to not use `unique_ptr` instead? I don't see sharing of message ownership.

File: src/net_message.h (hunk 1 above)
Comment (jonasschnelli) [reply src/net_message.h]: I kept it because it is moved code and to simplify the review.

Comment (promag) [reply src/net_message.h]: This is moved code right?

**Comment by jonasschnelli** [comment]
I think shared is fine here and does allow more flexible handling in future with little cost (ref counting)

**Comment by MarcoFalke** [comment]
utACK 59a7ce0f9de950b64c5e89dd29ebcd0b607b982e

**Comment by gmaxwell** [comment]
Adding an extra memory allocation for every network message isn't free. I'm concerned that we may be falling into overusing sharedptr "just in case", which is bad because it leaks performance in a diffuse way that doesn't show up well in profiles.  Please don't use sharedptr where unique will work.  If something new comes in in the future that needs a sharedptr then we can switch then.

**Comment by sipa** [comment]
@gmaxwell `unique_ptr` and `shared_ptr` have the same number of allocations (if `make_shared` is used, at least), though `shared_ptr` uses more memory (both for the pointer itself and for the allocated memory), and has atomics to update the refcounts (which may reduce performance).

File: src/net.cpp [hunk 2]
```diff
@@ -748,41 +748,40 @@ bool CNode::ReceiveMsgBytes(const char *pch, unsigned int nBytes, bool& complete
...
         // get current incomplete message, or create a new one
-        if (vRecvMsg.empty() ||
-            vRecvMsg.back().complete())
-            vRecvMsg.push_back(CNetMessage(Params().MessageStart(), SER_NETWORK, INIT_PROTO_VERSION));
+        if (vRecvMsg.empty() || vRecvMsg.back()->complete()) {
+            vRecvMsg.emplace_back(std::make_shared<CNetMessage>(Params().MessageStart(), SER_NETWORK, INIT_PROTO_VERSION));
+        }
 
-        CNetMessage& msg = vRecvMsg.back();
+        CNetMessageRef msg = vRecvMsg.back();
```
Comment (MarcoFalke) [inline src/net.cpp]: nit: In commit " Use shared pointers for CNetMessage handling " you could keep this a plain reference to keep the diff smaller, I think.

File: src/net_processing.cpp [hunk 3]
```diff
@@ -3002,29 +3002,29 @@ bool PeerLogicValidation::ProcessMessages(CNode* pfrom, std::atomic<bool>& inter
...
             return false;
         // Just take one message
         msgs.splice(msgs.begin(), pfrom->vProcessMsg, pfrom->vProcessMsg.begin());
-        pfrom->nProcessQueueSize -= msgs.front().vRecv.size() + CMessageHeader::HEADER_SIZE;
+        pfrom->nProcessQueueSize -= msgs.front()->vRecv.size() + CMessageHeader::HEADER_SIZE;
         pfrom->fPauseRecv = pfrom->nProcessQueueSize > connman->GetReceiveFloodSize();
         fMoreWork = !pfrom->vProcessMsg.empty();
     }
-    CNetMessage& msg(msgs.front());
+    CNetMessageRef msg(msgs.front());
```
Comment (MarcoFalke) [inline src/net_processing.cpp]: Same here

File: src/net_message.h [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    int readHeader(const char* pch, unsigned int nBytes);
+    int readData(const char* pch, unsigned int nBytes);
+
+    bool verifyMessageStart() const;
+    bool verifyHeader() const;
+    bool verifyChecksum(std::string& error) const;
+};
+
+typedef std::shared_ptr<NetMessageBase> NetMessageBaseRef;
+typedef std::shared_ptr<NetMessage> NetMessageRef;
```
Comment (251Labs) [inline src/net_message.h]: `NetMessageRef` doesn't seem to be used. Can `NetMessageRef` be deleted?

File: src/net_message.h [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+#include <protocol.h>
+#include <streams.h>
+#include <uint256.h>
+
+/** Maximum length of incoming protocol messages (no message over 4 MB is currently acceptable). */
+static const unsigned int MAX_PROTOCOL_MESSAGE_LENGTH = 4 * 1000 * 1000;
+
+// base class for format agnostic network messages
+class NetMessageBase
+{
```
[the remark this answers is not in the document]
Comment (jonasschnelli) [reply src/net_message.h]: Done

File: src/net.h [hunk 6]
```diff
@@ -567,9 +567,40 @@ class CNodeStats
...
+// base class for format agnostic network messages
+class NetMessageBase
+{
+public:
+    CDataStream vRecv; // received message data
+    int64_t nTime;     // time (in microseconds) of message receipt.
+
+    NetMessageBase(int nTypeIn, int nVersionIn) : vRecv(nTypeIn, nVersionIn)
+    {
+        nTime = 0;
```
Comment (promag) [inline src/net.h]: could move to init list `, nTime(0)`.

**Review by promag — COMMENTED** [review]
Fix commit message "Use shared pointers for CNetMessage handling".

**Comment by jonasschnelli** [comment]
Fixed commit message issue reported by @promag

File: src/net.h [hunk 7]
```diff
@@ -681,6 +633,7 @@ class CNode
 
     mapMsgCmdSize mapSendBytesPerMsgCmd;
     mapMsgCmdSize mapRecvBytesPerMsgCmd;
+    void RecordRecvBytesPerMsgCmd(const std::string& cmd, uint32_t bytes);
```
Comment (Empact) [inline src/net.h]: nit: could be private

File: src/net_message.h [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    {
+        if (!in_data) {
+            return 0;
+        }
+        return hdr.nMessageSize;
+    }
+
+    uint32_t GetMessageSizeWithHeader() const override
+    {
+        return hdr.nMessageSize + CMessageHeader::HEADER_SIZE;
```
Comment (Empact) [inline src/net_message.h]: Seems these other `hdr` accesses be guarded against `!in_data`, as they're uninitialized data in that case, in this case, `nMessageSize` would be `-1`.

File: src/net_message.cpp [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+{
+    assert(Complete());
+    if (data_hash.IsNull())
+        hasher.Finalize(data_hash.begin());
+    return data_hash;
+}
+
+bool NetMessage::VerifyMessageStart() const
+{
+    return memcmp(hdr.pchMessageStart, Params().MessageStart(), CMessageHeader::MESSAGE_START_SIZE) == 0 ? true : false;
```
Comment (practicalswift) [inline src/net_message.cpp]:
```
src/net_message.cpp:78:108: warning: redundant boolean literal in ternary expression result [readability-simplify-boolean-expr]
```

File: src/net_message.cpp [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+{
+    unsigned int nRemaining = hdr.nMessageSize - nDataPos;
+    unsigned int nCopy = std::min(nRemaining, nBytes);
+
+    if (vRecv.size() < nDataPos + nCopy) {
+        // Allocate up to 256 KiB ahead, but never more than the total message size.
+        vRecv.resize(std::min(hdr.nMessageSize, nDataPos + nCopy + 256 * 1024));
+    }
+
+    hasher.Write((const unsigned char*)pch, nCopy);
```
Comment (practicalswift) [inline src/net_message.cpp]:
```
src/net_message.cpp:61:18: warning: C-style casts are discouraged; use reinterpret_cast [google-readability-casting]
src/net_message.cpp:61:18: warning: do not use C-style cast to convert between unrelated types [cppcoreguidelines-pro-type-cstyle-cast]
```

File: src/net_message.h (hunk 8 above)
Comment (jonasschnelli) [reply src/net_message.h]: I basically moved this code part... I think its fine.

File: src/net_message.cpp (hunk 10 above)
Comment (jonasschnelli) [reply src/net_message.cpp]: It's moved code and the scope of this PR is not refactor.

File: src/net_message.cpp (hunk 9 above)
Comment (jonasschnelli) [reply src/net_message.cpp]: Fixed.

File: src/net.h (hunk 7 above)
Comment (jonasschnelli) [reply src/net.h]: Fixed.

**Review by ajtowns — COMMENTED** [review]
c2f9c04583a3ffcae4a54d85bb49def58e79cef2 looks reasonable to me

File: src/net_message.h [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+    int64_t nTime;     // time (in microseconds) of message receipt.
+
+    NetMessageBase(int nTypeIn, int nVersionIn) : vRecv(nTypeIn, nVersionIn)
+    {
+        nTime = 0;
+    }
+    virtual ~NetMessageBase() {}
+
+    virtual bool Complete() const = 0;
+    virtual uint32_t GetMessageSize() const = 0;           //returns 0 when message has not yet been parsed
```
Comment (practicalswift) [inline src/net_message.h]:
```
2018-09-22 21:22:44 cpplint(pr=14046): src/net_message.h:30:  Should have a space between // and comment  [whitespace/comments] [4]
```

File: src/net_message.h [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+    virtual uint32_t GetMessageSizeWithHeader() const = 0; //return complete message size including header
```
Comment (practicalswift) [inline src/net_message.h]:
```
2018-09-22 21:22:44 cpplint(pr=14046): src/net_message.h:31:  Should have a space between // and comment  [whitespace/comments] [4]
```

File: src/net_message.h [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+
+    virtual std::string GetCommandName() const = 0; //returns the command name. Returns an empty string when no command name is supported
```
Comment (practicalswift) [inline src/net_message.h]:
```
2018-09-22 21:22:44 cpplint(pr=14046): src/net_message.h:33:  Should have a space between // and comment  [whitespace/comments] [4]
```

File: src/net_message.h [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+
+    virtual void SetVersion(int nVersionIn)
+    {
+        vRecv.SetVersion(nVersionIn);
+    }
+
+    virtual int Read(const char* pch, unsigned int nBytes) = 0; //parse bytes
```
Comment (practicalswift) [inline src/net_message.h]:
```
2018-09-22 21:22:44 cpplint(pr=14046): src/net_message.h:40:  Should have a space between // and comment  [whitespace/comments] [4]
```

File: src/net_message.h [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+
+    virtual bool VerifyMessageStart() const = 0;
+    virtual bool VerifyHeader() const = 0;
+    virtual bool VerifyChecksum(std::string& error) const = 0;
+};
+
+//basic network message for the currently used unencryted p2p communication
```
Comment (practicalswift) [inline src/net_message.h]:
```
2018-09-22 21:22:44 cpplint(pr=14046): src/net_message.h:47:  Should have a space between // and comment  [whitespace/comments] [4]
```

File: src/net_message.cpp [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+    // copy data to temporary parsing buffer
+    unsigned int nRemaining = 24 - nHdrPos;
+    unsigned int nCopy = std::min(nRemaining, nBytes);
+
+    memcpy(&hdrbuf[nHdrPos], pch, nCopy);
+    nHdrPos += nCopy;
+
+    // if header incomplete, exit
+    if (nHdrPos < 24)
+        return nCopy;
```
Comment (practicalswift) [inline src/net_message.cpp]:
```
2018-09-25 21:41:13 clang(pr=14046): net_message.cpp:31:16: warning: implicit conversion changes signedness: 'unsigned int' to 'int' [-Wsign-conversion]
```

File: src/net_message.cpp [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    // reject messages larger than MAX_SIZE
+    if (hdr.nMessageSize > MAX_SIZE)
+        return -1;
+
+    // switch state to reading message data
+    in_data = true;
+
+    return nCopy;
```
Comment (practicalswift) [inline src/net_message.cpp]:
```
2018-09-25 21:41:13 clang(pr=14046): net_message.cpp:48:12: warning: implicit conversion changes signedness: 'unsigned int' to 'int' [-Wsign-conversion]
```

File: src/net_message.h [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+
+/** Maximum length of incoming protocol messages (no message over 4 MB is currently acceptable). */
+static const unsigned int MAX_PROTOCOL_MESSAGE_LENGTH = 4 * 1000 * 1000;
+
+// base class for format agnostic network messages
+class NetMessageBase
+{
+public:
+    CDataStream vRecv; // received message data
+    int64_t nTime;     // time (in microseconds) of message receipt.
```
Comment (promag) [inline src/net_message.h]: nit, `int64_t nTime{0};` and remove from the constructor.

**Review by promag — COMMENTED** [review]
utACK c2f9c04.

File: src/net_message.h (hunk 18 above)
Comment (251Labs) [reply src/net_message.h]: That's a nice nit @promag. Alternatively, `nTime` can be initialized with an empty initializer list `int64_t nTime {};`

**Comment by etscrivner** [comment]
utACK c2f9c04583a3ffcae4a54d85bb49def58e79cef2

File: src/net_message.cpp [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+
+int NetMessage::Read(const char* pch, unsigned int nBytes)
+{
+    return in_data ? ReadData(pch, nBytes) : ReadHeader(pch, nBytes);
+}
+
+int NetMessage::ReadHeader(const char* pch, unsigned int nBytes)
+{
+    // copy data to temporary parsing buffer
+    unsigned int nRemaining = 24 - nHdrPos;
```
[the remark this answers is not in the document]
Comment (jonasschnelli) [reply src/net_message.cpp]: Agree. Though I'd say scope of the PR is not to refactor code parts that where moved 1:1. Lets do further refactors in another PR.

File: src/net_message.h (hunk 18 above)
Comment (jonasschnelli) [reply src/net_message.h]: Agree. Though I'd say scope of the PR is not to refactor code parts that where moved 1:1. Lets do further refactors in another PR.

Comment (promag) [reply src/net_message.h]: SGTM.

**Comment by etscrivner** [comment]
ACK c2f9c04583a3ffcae4a54d85bb49def58e79cef2

File: src/net_message.h (hunk 15 above)
[the remark this answers is not in the document]
Comment (jonasschnelli) [reply src/net_message.h]: Fixed.

**Comment by jonasschnelli** [comment]
Rebased.

**Comment by promag** [comment]
utACK c2f9c04.

**Comment by sipa** [comment]
Is it really necessary to make `CNetMessage` a virtual class? This seems very much overkill. Only the network deserializers should change based on what protocol is used; we don't need to have that variability put a burden on every separate message.

**Comment by MarcoFalke** [comment]
@jonasschnelli Mind to follow up on @sipa's question? I have removed this from high priority for now, but happy to add back when you addressed the question.

**Comment by jonasschnelli** [comment]
Rebased

**Comment by sipa** [comment]
Approach NACK. This is overkill; the message parsing/serializing logic can just move out of CNetMessage to avoid turning it into a virtual class.

**Comment by jonasschnelli** [comment]
@sipa: Thanks for the review. Currently turning this into the approach you mentioned. I think it's doable though not super convinced right now due to the state necessary in parsing/serialization (hash, MAC).
Until I'm finished I keep this rebased to not lose testability of other elements in the pipeline.

**Comment by jonasschnelli** [comment]
#16202 seems to a superior solution. Will close this PR as soon as #14032 is rebased on top of #16202.

**Comment by jonasschnelli** [comment]
Superseded. Closing.

**Pull request closed without merging**